### PR TITLE
Faster search indexing

### PR DESCRIPTION
**Pull request by mmahalwy**
Rearranged the search to allow for better indexing and better/flexible search

### CODE REVIEW

File: app/controllers/concerns/search/language_detection.rb [hunk 1]
```diff
@@ -32,7 +32,9 @@ def initialize(headers, session, query)
       # Rails.logger.info "params #{ ap params }"
 
       boost_language_code = Hash.new
-      @indices_boost = Hash.new
+      @indices_boost = {
+        transliteration: 3
```
Comment (mmahalwy) [inline app/controllers/concerns/search/language_detection.rb]: give it a head start

File: app/models/content/translation.rb [hunk 2]
```diff
@@ -9,88 +9,85 @@
...
+    if codes.empty?
+      language_codes_array = Content::Resource.select(:language_code).distinct.map { |row| row.language_code }
+    else
+      language_codes_array = Content::Resource.where(language: codes).select(:language_code ).distinct.map { |row| row.language_code }
+    end
 
-                Rails.logger.debug( "importing index #{index_name_lc}" )
-                self.importing options
-            end
+    Parallel.each(language_codes_array, in_processes: 16, progress: 'Importing translations') do |language_code|
```
Comment (mmahalwy) [inline app/models/content/translation.rb]: this is so fast it might actually bring down your db server

Comment (mmahalwy) [inline app/models/content/translation.rb]: @ahmedre lol

File: app/models/search/query.rb [hunk 3]
```diff
@@ -170,10 +192,55 @@ def query_object
...
 
-        query[:bool][:must].unshift(terms) if hits_query?
+        bool[:must].unshift(terms) if hits_query?
+
+        bool
+      end
 
+      def dis_max_query
+        {
+          dis_max: {
```
Comment (mmahalwy) [inline app/models/search/query.rb]: I want to experiment with this more. It's really cool

File: config/elasticsearch/settings.yml [hunk 4]
```diff
@@ -60,6 +60,14 @@ index:
             quran_token_to_root_stop:
                 type: stop
                 stopwords_path: analysis/stop.text-token.to.text-root.txt
+            dbl_metaphone:
+                type: phonetic
```
Comment (sharabash) [inline config/elasticsearch/settings.yml]: > Undumpable Exception -- #<Elasticsearch::Transport::Transport::Errors::BadRequest: [400] {"error":"IndexCreationException[[transliteration] failed to create index]; nested: ElasticsearchIllegalArgumentException[failed to find token filter type [phonetic] for [dbl_metaphone]]; nested: NoClassSettingsException[Failed to load class setting [type] with value [phonetic]]; nested: ClassNotFoundException[org.elasticsearch.index.analysis.phonetic.PhoneticTokenFilterFactory]; ","status":400}>

Comment (sharabash) [inline config/elasticsearch/settings.yml]: Undumpable Exception -- #Elasticsearch::Transport::Transport::Errors::BadRequest: [400] {"error":"IndexCreationException[[transliteration] failed to create index]; nested: ElasticsearchIllegalArgumentException[failed to find token filter type [phonetic] for [dbl_metaphone]]; nested: NoClassSettingsException[Failed to load class setting [type] with value [phonetic]]; nested: ClassNotFoundException[org.elasticsearch.index.analysis.phonetic.PhoneticTokenFilterFactory]; ","status":400}

Comment (sharabash) [inline config/elasticsearch/settings.yml]: installing the elasticsearch-analysis-plugin is problematic. i can't find a ready URL (github repo won't work, elasticsearch complains that the plugin isn't built). building it with maven. if this doesn't work i'll give up for now

**Comment by sharabash** [comment]
bug: translation settings and mappings don't get applied; i think we need to manually put those

File: config/elasticsearch/settings.yml [hunk 5]
```diff
@@ -60,6 +60,14 @@ index:
...
+                encoder: double_metaphone
```
Comment (sharabash) [inline config/elasticsearch/settings.yml]: this should be `doublemetaphone`

File: lib/tasks/elasticsearch.rake [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+    @models.each do |model|
+      model = Kernel.const_get(model)
+      model.delete_index
+    end
+  end
+
+  desc 'setup all elasticsearch indices'
+  task setup_indices: :environment do
+    ActiveRecord::Base.logger = Logger.new(STDOUT)
+    Parallel.each(@models, in_processes: 16, progress: 'Importing models') do |model|
```
Comment (sharabash) [inline lib/tasks/elasticsearch.rake]: you're already running parallel in translations, i don't know if the processes multiply but a value of `16` here freezes my system in the middle. please reduce this to maybe `4`.

File: config/elasticsearch/settings.yml (hunk 5 above)
Comment (mmahalwy) [inline config/elasticsearch/settings.yml]: Hmmm.. are you sure? https://www.elastic.co/guide/en/elasticsearch/guide/current/phonetic-matching.html

**Comment by sharabash** [comment]
add mappings to text class, take out parallel or reduce it by a lot (optional, i can just change it temporarily too), change double_metaphone to doublemetaphone

then let me recreate indices and test this out

**Comment by sharabash** [comment]
it seems you've done that, let me try this from scratch

File: app/models/search/client.rb [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+          # fields: fields,
+          _source: source,
+          query: query_object
+        }
+      }
+    end
+
+    def request
+      @start_time = Time.now
+      binding.pry
```
Comment (sharabash) [inline app/models/search/client.rb]: this definitely needs to be commented out or removed O_O

**Comment by sharabash** [comment]
this is fine now, i think we can merge it. advice for future of search is that we limit our results to 5 pages or so... also there has to be a way to make it more lightweight, maybe i'll work on that soon. i.e. we want to know how long the actual query takes and how long the selecting of data takes, etc.

**Comment by sharabash** [comment]
also, i think probably search results should be in plain utf8 text and not have audio bound to it.
also, i think by default we need to pass in a language parameter, i.e. whatever the language is belonging to the selected translation

**Comment by sharabash** [comment]
merge ok :+1:

**Comment by sharabash** [comment]
need some patches, i'll comment in code

File: app/models/quran/text.rb [hunk 8]
```diff
@@ -9,33 +9,63 @@
...
+        similarity: 'my_bm25',
+        analyzer: 'autocomplete_arabic',
+        search_analyzer: 'arabic_normalized',
+        index_options: 'offsets'
+    end
+  end
+
+  def self.import( options = {} )
+    Quran::Text.connection.cache do
+      transform = lambda do |a|
```
Comment (sharabash) [inline app/models/quran/text.rb]: fix this so that it's like translations (no funky transforms)

File: app/models/content/translation.rb [hunk 9]
```diff
@@ -9,88 +9,88 @@
...
+      )
+    )
+  )
+
+  mappings _all: { enabled: false } do
+    indexes :text, type: 'multi_field' do
+      indexes :text, type: 'string', similarity: 'my_bm25',
+                     term_vector: 'with_positions_offsets_payloads',
+                     analyzer: 'standard'
+      indexes :stemmed, type: 'string', similarity: 'my_bm25',
```
Comment (sharabash) [inline app/models/content/translation.rb]: add this
```
      indexes :autocomplete, type: 'string', analyzer: 'autocomplete',
          search_analyzer: 'standard', index_options: 'offsets'
```

File: app/models/content/tafsir_ayah.rb [hunk 10]
```diff
@@ -5,46 +5,57 @@
...
+                        term_vector: 'with_positions_offsets_payloads',
+                        search_analyzer: 'arabic_normalized',
+                        analyzer: 'arabic_ngram'
     end
+  end
+
+  def as_indexed_json(options = {})
+    resource = tafsir.resource
+
+    as_json(include: :tafsir).merge(resource: resource.as_json(include: :language))
```
Comment (sharabash) [inline app/models/content/tafsir_ayah.rb]: also put `tafsir.text` in the base of the object and delete it out of the tafsir object

**Comment by sharabash** [comment]
ok, so this branch needs all of the patches delineated in the line comments and two modifications to `search/suggest.rb`:

delete lines 54-56:
```
          #if @lang == 'en'
          #  indices.push( 'translation' )
          #end
```

then change lines 103-104:

```
          ayah = "#{hit['_source']['ayah_key'].gsub(/_/,':')}"
          href = "/#{hit['_source']['ayah_key'].gsub(/_/,'/')}"
```

File: app/models/quran/text.rb (hunk 8 above)
Comment (mmahalwy) [inline app/models/quran/text.rb]: i dont understand :(

File: app/models/content/tafsir_ayah.rb (hunk 10 above)
Comment (mmahalwy) [inline app/models/content/tafsir_ayah.rb]: I dont understandy too :(

**Comment by mmahalwy** [comment]
@sharabash want to jump on this PR to help with the comments?

**Pull request merged**